library/src/main/java/com/finecinnamon/kats/Option.kt

```diff
@@ -0,0 +1,159 @@
+     * return an $option (which could be $none).
+     *
+     * @param  f   the function to apply
+     * @see map
+     * @see foreach
+     */
+    inline fun <B> flatMap(f: (A) -> Option<B>): Option<B> = if (isEmpty) None() else f(get())
+
+    /**
+     * Returns the result of applying $f to this $option's
+     * value if the $option is nonempty.  Otherwise, evaluates
+     * expression `ifEmpty`.
+     *
+     * @note This is equivalent to `$option map f getOrElse ifEmpty`.
+     *
+     * @param  ifEmpty the expression to evaluate if empty.
+     * @param  f       the function to apply if nonempty.
+     */
+    inline fun <B> fold(ifEmpty: () -> B, f: (A) -> B): B = if (isEmpty) ifEmpty() else f(get())
+
+    /**
+     * Returns this $option if it is nonempty '''and''' applying the predicate $p to
+     * this $option's value returns true. Otherwise, return $none.
+     *
+     *  @param  p   the predicate used for testing.
+     */
+    inline fun filter(p: (A) -> Boolean): Option<A> = if (isEmpty || p(get())) this else None()
+
+    /**
+     * Returns this $option if it is nonempty '''and''' applying the predicate $p to
+     * this $option's value returns false. Otherwise, return $none.
+     *
+     * @param  p   the predicate used for testing.
+     */
+    inline fun filterNot(p: (A) -> Boolean): Option<A> = if (isEmpty || !p(get())) this else None()
+
+    /**
+     * Returns false if the option is $none, true otherwise.
+     * @note   Implemented here to avoid the implicit conversion to Iterable.
+     */
+    val nonEmpty = isDefined
+
+    /**
+     * Tests whether the option contains a given value as an element.
+     *
+     * @param elem the element to test.
+     * @return `true` if the option has an element that is equal (as
+     * determined by `==`) to `elem`, `false` otherwise.
+     */
+    fun contains(elem: A): Boolean = !isEmpty && get() == elem
+
+    /**
+     * Returns true if this option is nonempty '''and''' the predicate
+     * $p returns true when applied to this $option's value.
+     * Otherwise, returns false.
+     *
+     * @param  p   the predicate to test
+     */
+    inline fun exists(p: (A) -> Boolean): Boolean = !isEmpty && p(get())
+
+    /**
+     * Returns true if this option is empty '''or''' the predicate
+     * $p returns true when applied to this $option's value.
+     *
+     * @param  p   the predicate to test
+     */
+    inline fun forall(p: (A) -> Boolean): Boolean = exists(p)
+
+    /**
+     * Apply the given procedure $f to the option's value,
+     * if it is nonempty. Otherwise, do nothing.
+     *
+     * @param  f   the procedure to apply.
+     * @see map
+     * @see flatMap
+     */
+    inline fun <B> foreach(f: (A) -> B) { if (!isEmpty) f(get()) }
+
+    class Some<A>(val value: A) : Option<A>() {
+        override val isEmpty = false
+        override fun get(): A = value
+    }
+
+    class None<A> : Option<A>() {
```

Comment:
Can this be an object instead of class so it's more effitient. `None` is a singleton that extends `Option<Nothing>`